File: backend/08-review/src/main/kotlin/com/dokandar/review/messaging/OutboxRelay.kt
package com.dokandar.review.messaging

import co.elastic.apm.api.ElasticApm
import com.dokandar.review.Config
import com.dokandar.review.data.Db
import com.dokandar.review.observability.Log
import com.dokandar.review.observability.Metrics
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.ByteArraySerializer
import org.apache.kafka.common.serialization.StringSerializer
import java.util.Properties
import kotlin.concurrent.thread

// Transactional-outbox relay: poll unsent (FOR UPDATE SKIP LOCKED), publish (acks=all, idempotent),
// mark sent. Publish-before-mark ⇒ at-least-once. Adaptive idle backoff.
object OutboxRelay {
    fun start() {
        if (Config.kafkaBootstrap.isEmpty()) return
        thread(isDaemon = true, name = "outbox-relay") {
            val props = Properties().apply {
                put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, Config.kafkaBootstrap)
                put(ProducerConfig.ACKS_CONFIG, "all"); put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true)
                put(ProducerConfig.CLIENT_ID_CONFIG, "08-review-outbox")
                put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer::class.java.name)
                put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer::class.java.name)
            }
            var producer: KafkaProducer<String, ByteArray>? = null
            while (producer == null) {
                try { producer = KafkaProducer(props); Log.info("review.outbox", "kafka producer connected") }
                catch (e: Exception) { Log.warn("review.outbox", "connect retry: ${e.message}"); Thread.sleep(5000) }
            }
            var idle = 0
            while (true) {
                val n = try { tick(producer) } catch (e: Exception) { Log.warn("review.outbox", "tick error: ${e.message}"); Thread.sleep(2000); continue }
                idle = if (n > 0) 0 else minOf(idle + 1, 5)
                Thread.sleep(1000L * (1 + idle))
            }
        }
    }

    private fun tick(producer: KafkaProducer<String, ByteArray>): Int = Db.tx { c ->
        val rs = c.prepareStatement("SELECT id,topic,key,payload FROM outbox WHERE sent_at IS NULL ORDER BY id LIMIT 100 FOR UPDATE SKIP LOCKED").executeQuery()
        data class Row(val id: Long, val topic: String, val key: String?, val payload: String)
        val rows = ArrayList<Row>(); while (rs.next()) rows.add(Row(rs.getLong("id"), rs.getString("topic"), rs.getString("key"), rs.getString("payload")))
        if (rows.isEmpty()) return@tx 0
        // The relay runs on a daemon thread with no APM transaction, so the kafka producer.send
        // would be an orphan span (dropped). Start + ACTIVATE a transaction so the producer spans
        // attach → Kafka shows in Dependencies + the service map, and the relay logs correlate.
        val tx = ElasticApm.startTransaction().apply { setType("messaging"); setName("OutboxRelay tick") }
        val scope = tx.activate()
        try {
            val sent = ArrayList<Long>()
            for (r in rows) {
                try { producer.send(ProducerRecord(r.topic, r.key, r.payload.toByteArray())).get(); sent.add(r.id) }
                catch (e: Exception) { Log.warn("review.outbox", "produce failed id=${r.id}: ${e.message}"); break }
            }
            if (sent.isNotEmpty()) {
                val arr = c.createArrayOf("bigint", sent.toTypedArray())
                c.prepareStatement("UPDATE outbox SET sent_at=now() WHERE id=ANY(?)").apply { setArray(1, arr) }.executeUpdate()
                Metrics.outboxPublished.labels(Metrics.SVC).inc(sent.size.toDouble())
                Log.info("review.outbox", "published ${sent.size} event(s)")
            }
            return@tx sent.size
        } finally {
            scope.close(); tx.end()
        }
    }
}

File: backend/07-coupon/src/Messaging/OutboxRelay.cs
using System.Text;
using Confluent.Kafka;
using Microsoft.EntityFrameworkCore;
using Elastic.Apm;
using Elastic.Apm.Api;
using Coupon.Data;
using Coupon.Observability;

namespace Coupon.Messaging;

// Transactional-outbox relay: poll unsent rows (FOR UPDATE SKIP LOCKED), publish to Kafka (acks=all,
// idempotent), mark sent_at. Publish-before-mark ⇒ at-least-once. Wrapped so it never blocks Kestrel.
public class OutboxRelay(IServiceProvider sp) : BackgroundService
{
    private IProducer<string, byte[]>? _producer;
    private int _idle;

    protected override async Task ExecuteAsync(CancellationToken ct)
    {
        await Task.Yield(); // return to host immediately — never block startup
        while (!ct.IsCancellationRequested && _producer == null)
        {
            try
            {
                var cfg = new ProducerConfig { BootstrapServers = Config.KafkaBootstrap, Acks = Acks.All, EnableIdempotence = true, ClientId = "07-coupon-outbox", MessageTimeoutMs = 10000 };
                _producer = new ProducerBuilder<string, byte[]>(cfg).Build();
                Log.Info("coupon.outbox", "kafka producer connected");
            }
            catch (Exception e) { Log.Warn("coupon.outbox", $"kafka connect retry: {e.Message}"); await Task.Delay(5000, ct); }
        }
        while (!ct.IsCancellationRequested)
        {
            int n;
            try { n = await TickAsync(ct); }
            catch (Exception e) { Log.Warn("coupon.outbox", $"tick error: {e.Message}"); await Task.Delay(2000, ct); continue; }
            _idle = n > 0 ? 0 : Math.Min(_idle + 1, 5); // back off when idle: 1s..6s (cuts idle DB polls + APM noise)
            await Task.Delay(TimeSpan.FromSeconds(Config.OutboxPollSeconds * (1 + _idle)), ct);
        }
    }

    private async Task<int> TickAsync(CancellationToken ct)
    {
        // Wrap the relay batch in an APM transaction so this BACKGROUND JOB is visible in APM
        // (the EF poll span + the Kafka producer spans nest under it). The .NET agent does NOT
        // auto-instrument Confluent.Kafka, so each produce gets an explicit messaging span with a
        // friendly "kafka" destination → Kafka shows in Dependencies + the service map.
        var apmTx = Agent.Tracer.StartTransaction("OutboxRelay process", "messaging");
        try
        {
        using var scope = sp.CreateScope();
        var db = scope.ServiceProvider.GetRequiredService<CouponDbContext>();
        await using var tx = await db.Database.BeginTransactionAsync(ct);
        var rows = await db.Outbox.FromSqlRaw(
            $"SELECT * FROM outbox WHERE sent_at IS NULL ORDER BY id LIMIT {Config.OutboxBatch} FOR UPDATE SKIP LOCKED").ToListAsync(ct);
        if (rows.Count == 0) { await tx.RollbackAsync(ct); return 0; }
        var sent = new List<long>();
        foreach (var r in rows)
        {
            try
            {
                var span = apmTx.StartSpan($"Kafka SEND to {r.Topic}", "messaging", "kafka", "send");
                span.Context.Destination = new Destination { Service = new Destination.DestinationService { Resource = "kafka", Name = "kafka", Type = "messaging" } };
                try { await _producer!.ProduceAsync(r.Topic, new Message<string, byte[]> { Key = r.Key ?? "", Value = Encoding.UTF8.GetBytes(r.Payload) }, ct); }
                finally { span.End(); }
                sent.Add(r.Id);
            }
            catch (Exception e) { Log.Warn("coupon.outbox", $"produce failed id={r.Id}: {e.Message}"); break; } // preserve id order
        }
        if (sent.Count > 0)
        {
            await db.Database.ExecuteSqlRawAsync("UPDATE outbox SET sent_at = now() WHERE id = ANY({0})", new object[] { sent.ToArray() });
            await tx.CommitAsync(ct);
            Metrics.OutboxPublished.WithLabels(Metrics.Svc).Inc(sent.Count);
            Log.Info("coupon.outbox", $"published {sent.Count} event(s)");
            return sent.Count;
        }
        await tx.RollbackAsync(ct);
        return 0;
        }
        finally { apmTx.End(); }
    }
}
